### PR TITLE
Add HostDecoderRandomCrop

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

### CODE REVIEW

**Comment by jantonguirao** [comment]
Build #614079

File: dali/image/jpeg_handle.cc [hunk 1]
```diff
@@ -0,0 +1,184 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+
+Modifications Copyright 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (JanuszL) [inline dali/image/jpeg_handle.cc]:
```suggestion
Copyright 2019, NVIDIA CORPORATION. All rights reserved.
```

Comment (jantonguirao) [reply dali/image/jpeg_handle.cc]: Will do

**Comment by jantonguirao** [comment]
Build (with HostDecoder + RandomResizedCrop): #619143 
Build (with HostDecoderRandomCrop + Resize): #619146

File: dali/image/generic_image.cc [hunk 2]
```diff
@@ -25,12 +25,37 @@ GenericImage::GenericImage(const uint8_t *encoded_buffer, size_t length, DALIIma
...
+      cv::Rect roi(x, y, newW, newH);
+      decoded_image(roi).copyTo(decoded_image_roi);
+      decoded_image = decoded_image_roi;
+      W = decoded_image.cols;
+      H = decoded_image.rows;
+      DALI_ENFORCE(W == newW);
+      DALI_ENFORCE(H == newH);
+  }
 
   if (decoded_image.data == nullptr) {
```
Comment (JanuszL) [inline dali/image/generic_image.cc]: It should be moved above L40

File: dali/pipeline/operators/decoder/host_decoder_test.cc [hunk 3]
```diff
@@ -68,7 +66,7 @@ class HostDecodeTestPng : public HostDecodeTest<ImageType> {
 TYPED_TEST_CASE(HostDecodeTestPng, Types);
 
 TYPED_TEST(HostDecodeTestPng, TestPngDecode) {
-  this->RunTestDecode(t_pngImgType);
+  this->RunTestDecode(t_pngImgType, 0.7);
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/pipeline/operators/decoder/host_decoder_test.cc]: Could you confirm if we were really falling back to OpenCV while now new decoder handles them?
I think new decoder is only for jpegs, and TIFF and PNG is still supported only by OpenCV.

File: docs/examples/pytorch/resnet50/main.py [hunk 4]
```diff
@@ -80,18 +80,21 @@
...
+        self.dali_cpu = dali_cpu
         self.input = ops.FileReader(file_root=data_dir, shard_id=args.local_rank, num_shards=args.world_size, random_shuffle=True)
         #let user decide which pipeline works him bets for RN version he runs
-        if dali_cpu:
+        if self.dali_cpu:
             dali_device = "cpu"
-            self.decode = ops.HostDecoder(device=dali_device, output_type=types.RGB)
+            # self.decode = ops.HostDecoder(device=dali_device, output_type=types.RGB)
+            self.decode = ops.HostDecoderRandomCrop(device=dali_device, output_type=types.RGB)
+            self.resize = ops.Resize(resize_shorter=crop)
```
Comment (JanuszL) [inline docs/examples/pytorch/resnet50/main.py]: I think it is wrong. We should resize to (crop, crop), so
```
self.resize = ops.Resize(resize_x=crop, resize_y=crop)
```

File: docs/examples/pytorch/resnet50/main.py [hunk 5]
```diff
@@ -106,7 +109,10 @@ def define_graph(self):
         rng = self.coin()
         self.jpegs, self.labels = self.input(name="Reader")
         images = self.decode(self.jpegs)
-        images = self.rrc(images)
+        if self.dali_cpu:
```
Comment (JanuszL) [inline docs/examples/pytorch/resnet50/main.py]: Maybe it is better to name it `self.res` in both cases and get rid of that if here.

File: dali/pipeline/operators/decoder/host_decoder_test.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/pipeline/operators/decoder/host_decoder_test.cc]: - Yes, I will confirm that this was the case
- Yes the new decoder is only for jpegs but the tests are actually written in such a way that first we are reading JPEGS, then convert them to TIFF/PNG when doing the test

File: docs/examples/pytorch/resnet50/main.py (hunk 4 above)
Comment (jantonguirao) [reply docs/examples/pytorch/resnet50/main.py]: Right, I'll change that

**Comment by jantonguirao** [comment]
Build #619258

File: dali/image/generic_image.cc (hunk 2 above)
Comment (jantonguirao) [reply dali/image/generic_image.cc]: Will do

File: docs/examples/pytorch/resnet50/main.py (hunk 5 above)
Comment (jantonguirao) [reply docs/examples/pytorch/resnet50/main.py]: Will do

**Comment by jantonguirao** [comment]
Build #619262

**Comment by jantonguirao** [comment]
Build #620181

File: dali/pipeline/operators/decoder/host_decoder_test.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/pipeline/operators/decoder/host_decoder_test.cc]: @JanuszL I confirm that in master, there are two images in our unit test dataset that cannot be decoded with our TurboJPEG decoder and we fallback to OpenCV. 
The new decoder in this PR handles those images without problems.

**Comment by jantonguirao** [comment]
Builds #621087

File: cmake/Dependencies.cmake [hunk 6]
```diff
@@ -87,10 +87,10 @@ endif()
 # libjpeg-turbo
 ##################################################################
 if (BUILD_JPEG_TURBO)
-  find_package(JpegTurbo 1.5 REQUIRED)
-  include_directories(SYSTEM ${JpegTurbo_INCLUDE_DIR})
-  list(APPEND DALI_LIBS ${JpegTurbo_LIBRARY})
-  list(APPEND DALI_EXCLUDES libturbojpeg.a)
+  find_package(JPEG REQUIRED)
```
Comment (JanuszL) [inline cmake/Dependencies.cmake]: Don't we need to enforce any version?

Comment (JanuszL) [reply cmake/Dependencies.cmake]: How about using JPEG_TURBO_ROOT_DIR when user doesn't have JPEG installed, but have it compiled on the side?

File: cmake/modules/FindJpegTurbo.cmake [hunk 7]
```diff
@@ -1,50 +0,0 @@
...
-# The following variables are optionally searched for defaults
-#  JPEG_TURBO_ROOT_DIR:         Base directory where all JPEG_TURBO components are found
-#
-# The following are set after configuration is done:
-#  JpegTurbo_FOUND
-#  JpegTurbo_INCLUDE_DIR
-#  JpegTurbo_LIBRARY
-#  JpegTurbo_VERSION
-
-set(JPEG_TURBO_ROOT_DIR "" CACHE PATH "Folder contains JpegTurbo")
```
Comment (JanuszL) [inline cmake/modules/FindJpegTurbo.cmake]: We need to update docs removing info about JPEG_TURBO_ROOT_DIR

Comment (JanuszL) [reply cmake/modules/FindJpegTurbo.cmake]: I think this file should stay, but it should look for JPEG

Comment (jantonguirao) [reply cmake/modules/FindJpegTurbo.cmake]: about that, there is a native package to find jpeg:
https://cmake.org/cmake/help/v3.0/module/FindJPEG.html
Do you think we need a custom one?

Comment (JanuszL) [reply cmake/modules/FindJpegTurbo.cmake]: It definitely doesn't handle the situation when lib is custom build in some custom path. Or does it?

File: cmake/Dependencies.cmake (hunk 6 above)
Comment (jantonguirao) [reply cmake/Dependencies.cmake]: You can specify custom location like this:
cmake -DJPEG_INCLUDE_DIR=/home/janton/libjpeg-turbo-1.5.3/install/include/ -DJPEG_LIBRARY=/home/janton/libjpeg-turbo-1.5.3/install/lib/libjpeg.so ..

File: cmake/modules/FindJpegTurbo.cmake (hunk 7 above)
Comment (jantonguirao) [reply cmake/modules/FindJpegTurbo.cmake]: It does:
cmake -DJPEG_INCLUDE_DIR=/home/janton/libjpeg-turbo-1.5.3/install/include/ -DJPEG_LIBRARY=/home/janton/libjpeg-turbo-1.5.3/install/lib/libjpeg.so ..

File: cmake/Dependencies.cmake (hunk 6 above)
Comment (jantonguirao) [reply cmake/Dependencies.cmake]: Version enforced in last commit

**Comment by jantonguirao** [comment]
Build #622030

File: dali/util/random_crop_generator.cc [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  , uniform_(0.0f, 1.0f)
+  , rand_gen_(seed)
+  , seed_(seed) {
+  }
+
+CropWindow RandomCropGenerator::GenerateCropWindowImpl(int H, int W) {
+    DALI_ENFORCE(H > 0);
+    DALI_ENFORCE(W > 0);
+
+    constexpr std::size_t kMaxAttempts = 10;
```
Comment (Kh4L) [inline dali/util/random_crop_generator.cc]: Why this number?

File: dali/pipeline/operators/decoder/host_decoder.cc [hunk 9]
```diff
@@ -12,10 +12,43 @@
...
+#include <opencv2/opencv.hpp>
+#include <tuple>
+#include <memory>
+#include "dali/image/image_factory.h"
 #include "dali/pipeline/operators/decoder/host_decoder.h"
 
 namespace dali {
 
+void HostDecoder::RunImpl(SampleWorkspace *ws, const int idx) {
+  auto &input = ws->Input<CPUBackend>(idx);
```
Comment (Kh4L) [inline dali/pipeline/operators/decoder/host_decoder.cc]:
```suggestion
  const auto &input = ws->Input<CPUBackend>(idx);
```

File: dali/util/random_crop_generator.cc (hunk 8 above)
Comment (jantonguirao) [reply dali/util/random_crop_generator.cc]: it is the same as in RandomResizedCrop

File: dali/pipeline/operators/decoder/host_decoder.cc (hunk 9 above)
Comment (jantonguirao) [reply dali/pipeline/operators/decoder/host_decoder.cc]: done

File: docs/examples/pytorch/resnet50/main.py [hunk 10]
```diff
@@ -84,14 +84,16 @@ def __init__(self, batch_size, num_threads, device_id, data_dir, crop, dali_cpu=
         #let user decide which pipeline works him bets for RN version he runs
         if dali_cpu:
             dali_device = "cpu"
-            self.decode = ops.HostDecoder(device=dali_device, output_type=types.RGB)
+            # self.decode = ops.HostDecoder(device=dali_device, output_type=types.RGB)
```
Comment (JanuszL) [inline docs/examples/pytorch/resnet50/main.py]: Remove

Comment (JanuszL) [reply docs/examples/pytorch/resnet50/main.py]: I think TensorFlow RN50 example could be updated as well.

Comment (jantonguirao) [reply docs/examples/pytorch/resnet50/main.py]: Will do

**Comment by jantonguirao** [comment]
Build #622160

**Comment by jantonguirao** [comment]
Build #623189

**Pull request merged**